assignments/HW4.md

```diff
@@ -0,0 +1,194 @@
+
+Then, submit this tagged version and a pdf of your Swagger documentation that describes the API endpoints to the associated Gradescope assignment.
+You can create a pdf of your Swagger documentation with the Print Preview function of your browser.
+
+### (2) Microservice ML feature implementation - (due Tuesday, Nov 1st 11:59 PM) | 55 points
+
+We provide you with a starter repository (see above) with a (synthetic) historical dataset (as a csv file) of student performance (this is an augmented version of the dataset found here), and a (bad) model trained over it. Start by exploring the data, to understand what information is available. Student performance is described by their grades (G1, G2, and G3), where G3 is the final year grade, the target performance metric. Grades are on a scale of 0-20; 20 is the best. CMU considers a high-quality student one with a G3 grade of 15 or higher.
+
+The existing model in the repository is very simple. Once you have familiarized yourself with the dataset, explore this model using whatever means you think are appropriate and determine in which ways you don’t believe it to be suitable. Based on what you learn in this exploratory step, you should train a new model and build a deployable microservice.
+
+You should evaluate your new model, and should explain how this model (overall) performs better than the baseline model, and any ways in which your model performs worse than the baseline model.
+
+Plan some testing, perform that testing on your microservice endpoints, and document it. At least some of this testing should be automated using Github Actions.
+As you have in previous assignments, you can create a test suite (ie. [pytest](https://flask.palletsprojects.com/en/2.2.x/testing/)) and have your workflow run the test suite.
+
+The deliverable for this task is a documented technical artifact; submit a link of your github repository to the Gradescope assignment. This should include:
+
+1. a trained model that is deployable as a microservice; the microservice must take as input a query request with student data and return a prediction about that student’s graduate school success. This is the API for your microservice. It should also include
+
+2. an updated README that clearly and concisely explains
+
+   - Clear documentation for your API. (How it should be called, what data it expects, any pre-conditions for the service, and how to understand the output).
+
+   - A description of which features you used in training your model, and how your retrained the model performs better than the baseline model
+
+   - testing and deployment instructions; and
+
+   - an explanation and justification of the testing you have done on it.
+
+### (3) Design document - (due Thursday, Nov 3rd, 23:59) | 70 points
```

Comment:
Just my two cents: Since we're forcing a service-based approach, I think that we should completely change this activity to focus exclusively on the design and integration of the REST API.

@namdy0429 @MichaelHilton @rohanpadhye may have thoughts?